clang/test/CIR/CodeGen/no-prototype.c

```diff
@@ -0,0 +1,51 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
+
+// No-proto definition followed by a correct call.
+int noProto0(x) int x; { return x; }
+int test0(int x) {
+  // CHECK: cir.func @test0
+  return noProto0(x); // We know the definition. Should be a direct call.
+  // CHECK: %{{.+}} = cir.call @noProto0(%{{.+}})
+}
+
+// Declaration without prototype followed by its definition, then a correct call.
+//
+// Call to no-proto is made after definition, so a direct call can be used.
+int noProto1();
+int noProto1(int x) { return x; }
+int test1(int x) {
+  // CHECK: cir.func @test1
+  return noProto1(x);
+  // CHECK: %{{.+}} = cir.call @noProto1(%{{[0-9]+}})
+}
+
+// Declaration without prototype followed by a correct call, then its definition.
+//
+// Call to no-proto is made before definition, so a variadic call that takes anything
+// is created. Later, when the definition is found, no-proto is replaced.
+int noProto2();
+int test2(int x) {
+  return noProto2(x);
+  // CHECK: %[[#PTR:]] = cir.get_global @noProto2 : cir.ptr <!cir.func<!s32i (!s32i)>>
+  // CHECK: %[[#C_PTR:]] = cir.cast(bitcast, %[[#PTR]] : !cir.ptr<!cir.func<!s32i (!s32i)>>), !cir.ptr<!cir.func<!s32i (!s32i)>>
+  // CHECK: %{{.+}} = cir.call %[[#C_PTR]](%{{.+}}) : (!cir.ptr<!cir.func<!s32i (!s32i)>>, !s32i) -> !s32i
+}
+int noProto2(int x) { return x; }
+
+// No-proto declaration without definition (any call here is "correct").
+//
+// Call to no-proto is made before definition, so a variadic call that takes anything
+// is created. Definition is not in the translation unit, so it is left as is.
+int noProto3();
+// cir.func private @noProto3(...) -> !s32i
+int test3(int x) {
+// CHECK: cir.func @test3
+  return noProto3(x);
+  // %[[#PTR:]] = cir.get_global @noProto3 : cir.ptr <!cir.func<!s32i (...)>>
+  // %[[#C_PTR:]] = cir.cast(bitcast, %[[#PTR]] : !cir.ptr<!cir.func<!s32i (...)>>), !cir.ptr<!cir.func<!s32i (!s32i)>>
+  // %{{.+}} = cir.call %[[#C_PTR]](%%{{.+}}) : (!cir.ptr<!cir.func<!s32i (!s32i)>>, !s32i) -> !s32i
+}
+
+// TODO(cir): Handle incorrect calls to no-proto functions. It's mostly undefined
```

Comment:
How does LLVM codegen handles that right now? If we add the `#cir.no_prototype` like mentioned above, that will allow the verifiers to handwave on the mismatching and things would compile fine. Later passes could use that information to either take advantage for optimizations or provide UB related diagnostics.

Reply:
> How does LLVM codegen handles that right now?

Extra args are just dropped, but I'm not sure about type mismatches and missing args. I'll take a look at it later.

> Later passes could use that information to either take advantage for optimizations or provide UB related diagnostics

Neat!